src/Workspaces/Core/Portable/FlowAnalysis/SymbolUsageAnalysis/SymbolUsageAnalysis.BasicBlockAnalysisData.cs

```diff
@@ -0,0 +1,178 @@
+
+            private BasicBlockAnalysisData()
+            {
+                _reachingWrites = new Dictionary<ISymbol, PooledHashSet<IOperation>>();
+            }
+
+            public static BasicBlockAnalysisData GetInstance() => s_pool.Allocate();
+
+            public void Free()
+            {
+                FreeAndClearValues();
+                s_pool.Free(this);
+            }
+
+            private void FreeAndClearValues()
+            {
+                foreach (var value in _reachingWrites.Values)
+                {
+                    value.Free();
+                }
+
+                _reachingWrites.Clear();
+            }
+
+            public void SetAnalysisDataFrom(BasicBlockAnalysisData other)
+            {
+                if (ReferenceEquals(this, other))
+                {
+                    return;
+                }
+
+                FreeAndClearValues();
+                AddEntries(_reachingWrites, other);
+            }
+
+            /// <summary>
+            /// Gets the currently reachable writes for the given symbol.
+            /// </summary>
+            public IEnumerable<IOperation> GetCurrentWrites(ISymbol symbol)
+            {
+                if (_reachingWrites.TryGetValue(symbol, out var values))
+                {
+                    foreach (var value in values)
+                    {
+                        yield return value;
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Marks the given symbol write as a new unread write operation,
+            /// potentially clearing out the prior write operations if <paramref name="maybeWritten"/> is <code>false</code>.
+            /// </summary>
+            public void OnWriteReferenceFound(ISymbol symbol, IOperation operation, bool maybeWritten)
+            {
+                if (!_reachingWrites.TryGetValue(symbol, out var values))
+                {
+                    values = PooledHashSet<IOperation>.GetInstance();
+                    _reachingWrites.Add(symbol, values);
+                }
+                else if (!maybeWritten)
+                {
+                    values.Clear();
+                }
+
+                values.Add(operation);
+            }
+
+            public bool Equals(BasicBlockAnalysisData other)
+                => this.GetHashCode() == (other?.GetHashCode() ?? 0);
+
+            public override bool Equals(object obj)
+            {
+                return Equals(obj as BasicBlockAnalysisData);
+            }
+
+            public override int GetHashCode()
+            {
+                var hashCode = _reachingWrites.Count;
+                foreach ((int keyHash, int valueHash) in _reachingWrites.Select(GetKeyValueHashCodeTuple).OrderBy(hashTuple => hashTuple.keyHash))
+                {
+                    hashCode = Hash.Combine(Hash.Combine(keyHash, valueHash), hashCode);
+                }
```

Comment:
1. using linq in a hash-function is generally not advised. esp. given the efforts you've put in this type to prevent allocations.
2. if this is going to be called multiple times, consider caching hashCode.

Reply:
I ended up removing the GetHashCode override completely as this type did not need to be an `IEquatable`. It was not used as a key for any map, and the only place where equality is checked is when the dataflow walker explicitly calls to compare equality of given basic block data instances, and now we explicitly compare the contents of the maps in that callback.

---
In reply to: [231669995](https://github.com/dotnet/roslyn/pull/31025#discussion_r231669995) [](ancestors = 231669995)